rust/lance/src/io/exec/knn.rs

```diff
@@ -299,103 +298,41 @@ impl FilterLoader for SelectionVectorToPrefilter {
-                            "Failed to calculate KNN: {e}"
-                        ))))
-                        .await
-                        .expect("KNNIndex failed to send message");
-                        return;
-                    }
-                };
-
-                if !tx.is_closed() {
-                    if let Err(e) = tx.send(Ok(result)).await {
-                        eprintln!("KNNIndex tx.send error: {e}")
-                    };
-                }
-                drop(tx);
-            }
-            .in_current_span(),
-        );
-
-        Self {
-            rx,
-            bg_thread: Some(bg_thread),
-        }
-    }
-}
-
-impl DFRecordBatchStream for KNNIndexStream {
-    fn schema(&self) -> arrow_schema::SchemaRef {
-        Arc::new(Schema::new(vec![
-            Field::new(DIST_COL, DataType::Float32, true),
-            ROW_ID_FIELD.clone(),
-        ]))
-    }
-}
-
-impl Stream for KNNIndexStream {
-    type Item = std::result::Result<RecordBatch, datafusion::error::DataFusionError>;
-
-    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
-        let this = Pin::into_inner(self);
-        // We need to check the JoinHandle to make sure the thread hasn't panicked.
-        let bg_thread_completed = if let Some(bg_thread) = &mut this.bg_thread {
-            match bg_thread.poll_unpin(cx) {
-                Poll::Ready(Ok(())) => true,
-                Poll::Ready(Err(join_error)) => {
-                    return Poll::Ready(Some(Err(DataFusionError::Execution(format!(
-                        "ExecNode(KNNIndexStream): thread panicked: {}",
-                        join_error
-                    )))));
-                }
-                Poll::Pending => false,
-            }
-        } else {
-            false
-        };
-        if bg_thread_completed {
-            // Need to take it, since we aren't allowed to poll if again after.
-            this.bg_thread.take();
-        }
-        // this.rx.
-        this.rx.poll_recv(cx)
-    }
+fn knn_index_stream(
+    query: Query,
+    dataset: Arc<Dataset>,
+    index_meta: Vec<Index>,
+    allow_list_input: Option<Box<dyn FilterLoader>>,
+) -> impl DFRecordBatchStream {
+    let pre_filter = Arc::new(PreFilter::new(
+        dataset.clone(),
+        &index_meta,
+        allow_list_input,
+    ));
+
+    let schema = Arc::new(Schema::new(vec![
+        Field::new(DIST_COL, DataType::Float32, true),
+        ROW_ID_FIELD.clone(),
+    ]));
+
+    let s = stream::iter(index_meta)
+        .zip(stream::repeat((
+            dataset.clone(),
+            pre_filter.clone(),
+            query.clone(),
+        )))
+        .map(|(idx, (ds, pre_filter, query))| async move {
+            let index = ds
+                .open_vector_index(&query.column, &idx.uuid.to_string())
+                .await?;
+            index.search(&query, pre_filter.clone()).await
```

Comment:
should we put `open_vector_index` and `search` on difference threads?

Reply:
what is the difference? index has data dependency here. You mean like this?

```
indices = io_thread_pool.submits(ds.open_vector_index(...))
results = cpu_pool.submits(index.search(...))
```